Create dropdown that will list databases.

hilakey opened the pull request:
Here's how the database list will look on a live server: [](url) https://play-user-data-beetle.uc.r.appspot.com/
+ list of databases has been temporarily hard coded
+ Modify formatting
+Remove white spaces
+Rename files

hilakey committed:
Create dropdown that will list databases.

gagomez commented on src/main/java/com/google/sps/servlets/exampledbServlet.java:
```diff
@@ -0,0 +1,77 @@
+
+/** Servlet that returns HTML that contains content from the example database. */
+@WebServlet("/example")
+public class exampledbServlet extends HttpServlet {
```
Upper camelcase here and in the file name.

Also, is this an example or is this the actual get databases servlet?

hilakey replied:
This is the actual get servlet, but I just didn't give it a more conventional name at the time. But I just fixed it.

gagomez commented on src/main/java/com/google/sps/servlets/exampledbServlet.java:
```diff
@@ -0,0 +1,77 @@
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that returns HTML that contains content from the example database. */
```
This doesn't return HTML right? It returns the list of databases configured for beetle.

hilakey replied:
Originally it was HTML but it actually returns JSON data, so I updated the comment to reflect this!

gagomez commented on src/main/java/com/google/sps/servlets/exampledbServlet.java:
```diff
@@ -0,0 +1,77 @@
+    try (ResultSet resultSet =
+        dbClient
+            .singleUse() // Execute a single read or query against Cloud Spanner.
+            .executeQuery(Statement.of("SELECT option_name FROM information_schema.database_options WHERE catalog_name = '' and schema_name = ''"))) {
```
These are options set on a single database like permissions and such. I think what we want here is a list of databases right? We won't get that by querying a single database. For now, let's just have this be a hardcoded list of IDs, later we can look into using a configuration file to specify these.

hilakey replied:
All right, done. As I was looking for documentation on listing databases it seems cloud spanner has...very little information on this.

gagomez commented on src/main/webapp/index.html:
```diff
@@ -3,9 +3,18 @@
     <p>Click <a href="/example-db">here</a> to view data queried from an example Cloud Spanner database.</p>
+    <p>please select the database you would like to use</p>
+
+    <div id="select-db">
+        <select id="list-databases" value="list-databases" class="select">
+            <option id="select-title">Select a Database</option>
+        </select>
+    </div>
```
Should we start with a "Loading..." message until the list is retrieved?

sasymer replied:
I like that idea too!

hilakey replied:
Done!

gagomez commented on src/main/webapp/index.html:
```diff
@@ -3,9 +3,18 @@
+    <p>please select the database you would like to use</p>
+
+    <div id="select-db">
+        <select id="list-databases" value="list-databases" class="select">
```
What does setting the value to list-databases do? It isn't even an option in the select.

hilakey replied:
That is true, I guess I'm used to examples having values for many elements even if they aren't necessary, but I removed the value.

gagomez commented on src/main/java/com/google/sps/servlets/exampledbServlet.java:
```diff
@@ -0,0 +1,77 @@
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that returns HTML that contains content from the example database. */
+@WebServlet("/example")
```
Let's come up with a more descriptive path. For example, "/databases"

hilakey replied:
Done

gagomez commented on src/main/webapp/script.js:
```diff
@@ -0,0 +1,37 @@
+      console.log(list);
+
+      let dropdown = document.getElementById('list-databases');
+      dropdown.length = 0;
```
This seems odd. I am guessing you are doing this to clear the options set in the dropdown?

I would suggest `dropdown.options.remove(0)` as it is considerably easier to read.

hilakey replied:
Yes that's what it does, I wasn't familiar with this other option you showed but I do see how it's easier to read! Thanks for the tip just changed it!

gagomez commented on src/main/webapp/script.js:
```diff
@@ -0,0 +1,37 @@
+            }
+      	  
+        }
+
+
```
Let's remove this unnecessary whitespace as well as all these console logs which I imagine where for dev testing.

hilakey replied:
Done

gagomez reviewed (commented):
Remove all the additions from target/* as those are built from the source which is what we want to have in github.

sasymer commented on src/main/java/com/google/sps/servlets/exampledbServlet.java:
```diff
@@ -0,0 +1,77 @@
+            .singleUse() // Execute a single read or query against Cloud Spanner.
+            .executeQuery(Statement.of("SELECT option_name FROM information_schema.database_options WHERE catalog_name = '' and schema_name = ''"))) {
+       while(resultSet.next()){
+           /*response
+            .getWriter()
+            .println(
+                String.format(
+                    "<p>%s</p>",
+                    resultSet.isNull(0) ? "" : resultSet.getString(0)));*/
```
Maybe could get rid of this comment unless you're going to use it?

hilakey replied:
Done

sasymer commented on target/beetle-1/index.html:
```diff
@@ -0,0 +1,20 @@
+    <p>Click <a href="/example-db">here</a> to view data queried from an example Cloud Spanner database.</p>
+    <p>please select the database you would like to use</p>
+
+    <div id="select-db">
+        <select id="list-databases" value="list-databases" class="select">
+            <option id="select-title">Select a Database</option>
+        </select>
+    </div>
```
Indentation

sasymer commented on src/main/java/com/google/sps/servlets/exampledbServlet.java:
```diff
@@ -0,0 +1,77 @@
+  DatabaseClient dbClient;
+
+  public void init() {
+    Spanner spanner = SpannerOptions.newBuilder().build().getService();
+    DatabaseId db = DatabaseId.of("play-user-data-beetle", "test-instance" , "example-db");
+    this.dbClient = spanner.getDatabaseClient(db);
```
Eventually (when hard coding is taken out), probably won't need this part because I don't think you need to open specific databases. I'm not sure about this, though, because you use the dbClient below.

sasymer commented on src/main/java/com/google/sps/servlets/exampledbServlet.java:
```diff
@@ -0,0 +1,77 @@
+       }
+  }
+
+  private String convertToJsonUsingGson(List<String> listDatabases) {
+    Gson gson = new Gson();
+    String json = gson.toJson(listDatabases);
+    return json;
+  }
```
Nice helper method

sasymer reviewed (changes requested):
I think target files should be deleted before merging (per what Jiaxin said on my pull request)

hilakey committed:
Add temporary hardcoded db list.
Add "loading.." text until list has been filled.
Remove console.log() functions.
Remove whitespace.
Adjust formatting.

hilakey committed:
Remove target folder.
Add Filters.html file

wdcroslen reviewed (approved):
Looks good! I like the logic in your Javascript and how the databases are added to the dropdown!

jiaxinz-g commented on src/main/java/com/google/sps/servlets/BeetleDbServlet.java:
```diff
@@ -0,0 +1,64 @@
+@WebServlet("/databases")
+public class BeetleDbServlet extends HttpServlet {
+
+  DatabaseClient dbClient;
```
Since we are hardcoding the list of DBs to return, these are not needed. Let's remove them, so as the corresponding unused imports.

hilakey replied:
Done.

jiaxinz-g commented on src/main/webapp/Filters.html:
```diff
@@ -0,0 +1,6 @@
+<!DOCTYPE html>
```
Let's not add this file until it is being used, would be better to keep it in the local change if you plan to use it in a followup PR.

hilakey replied:
Ah okay just un-added it.

jiaxinz-g reviewed (approved):
Two minor comments. The rest looks good to me!

hilakey committed:
Remove database syntax.

hilakey committed:
Merge branch 'master' into MillenniaDev
